README.md

#Oxford Cryosystem 700 Tango DS

##Communication protocol:
http://connect.oxcryo.com/serialcomms/700series/cs_status.html#extended


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!--## [Unreleased] -->
<!--### Added -->

## [2.0.X] 
### Added
 - Migrate to Tango9 and python 3

### Removed
- Command CryoShutter_Start_Auto 
### Fixed



[keepachangelog.com]: http://keepachangelog.com
[2.0.X]: https://github.com/ALBA-Synchrotron/OxfCryo700/python2-eol...HEAD


oxfcryo700/tango.py
import threading
import serial
from tango import DevState
from tango.server import Device, attribute, command
from tango.server import device_property
from .oxfordcryo import StatusPacket, CSCOMMAND, splitBytes


class OxfCryo700(Device):
    port = device_property(dtype=str, doc='Serial port name (/dev/ttyXX)')

    def init_device(self):
        Device.init_device(self)
        self.info_stream('In Python init_device method')
        self.serial = serial.Serial(self.port)
        self.status_packet = None
        self.status_thread = threading.Thread(group=None,
                                              target=self.update_status_packet)
        self.status_thread_stop = threading.Event()
        self.status_thread.start()
        self.set_state(DevState.ON)

    def delete_device(self):
        self.status_thread_stop.set()
        self.status_thread.join(3.0)
        self.info_stream('OxfCryo700.delete_device')

    def _write(self, data):
        data_bytes = bytes(data)
        self.serial.write(data_bytes)
    # ------------------------------------------------------------------
    # COMMANDS
    # ------------------------------------------------------------------

    @command
    def Restart(self):
        data = [2, CSCOMMAND.RESTART]
        self.debug_stream("Restart(): sending data: {}".format(data))
        self._write(data)

    @command
    def Purge(self):
        data = [2, CSCOMMAND.PURGE]
        self.debug_stream("PURGE(): sending data: {}".format(data))
        self._write(data)

    @command
    def Stop(self):
        data = [2, CSCOMMAND.STOP]
        self.debug_stream("Stop(): sending data: {}".format(data))
        self._write(data)

    @command(dtype_in=(float,), doc_in='Rate and FinalTemperature')
    def Ramp(self, args):
        """
        The CSCOMMAND_RAMP command packet, size = 6
        The Params[] array consists of a short containing desired ramp rate
        in K/hour,
        followed by a short containing the end temperature in centi-Kelvin
        """

        if len(args) != 2:
            raise ValueError("Wrong number of arguments. Required paramters "
                             "are ramp and end temperature.")

        rate = int(args[0])
        finalTemp = args[1]

        finalTemp = int(finalTemp * 100)  # transfering to centi-Kelvin
        rateHigh, rateLow = splitBytes(rate)
        finalTempHigh, finalTempLow = splitBytes(finalTemp)
        data = [6, CSCOMMAND.RAMP, rateHigh, rateLow, finalTempHigh,
                finalTempLow]
        self.debug_stream("Ramp(): sending data: {}".format(data))
        self._write(data)

    @command(dtype_in=bool, doc_in='Turn on the Turbo')
    def Turbo(self, turn_on):
        """
        The CSCOMMAND_TURBO command packet, size = 3
        The Params[] array consists of a single char taking the value either 0
        (switch Turbo off) or 1 (switch Turbo on)
        """
        if turn_on:
            turboState = 1
        else:
            turboState = 0
        data = [3, CSCOMMAND.TURBO, turboState]
        self.debug_stream("Turbo(): sending data: {}".format(data))
        self._write(data)

    @command(dtype_in=float, doc_in='Temperature between 80 to 400 Kelvins')
    def Cool(self, temp):
        """
        The	CSCOMMAND_COOL command packet, size = 4
        The Params[] array consists of a short containing the end
        temperature in centi-Kelvin
        """
        if not (80 <= temp <= 400):
            raise ValueError("Wrong arguments. Cool only accept values "
                             "between 80 to 400")

        cool_value = int(temp * 100)
        HIBYTE, LOBYTE = splitBytes(cool_value)
        data = [4, CSCOMMAND.COOL, HIBYTE, LOBYTE]
        self.debug_stream("Cool(): sending data: {}".format(data))
        self._write(data)

    @command
    def Pause(self):
        data = [2, CSCOMMAND.PAUSE]
        self.debug_stream("Pause(): sending data:{}".format(data))
        self._write(data)

    @command
    def Resume(self):
        data = [2, CSCOMMAND.RESUME]
        self.debug_stream("Resume(): sending data: {}".format(data))
        self._write(data)

    @command(dtype_in=int, doc_in='Plat command identifier - parameter '
                                  'follows')
    def Plat(self, val):
        """
        The CSCOMMAND_PLAT command packet, size = 4
        The Params[] array consists of a short containing the duration of the
        Plat in minutes
        """

        HIBYTE, LOBYTE = splitBytes(val)
        data = [4, CSCOMMAND.PLAT, HIBYTE, LOBYTE]
        self.debug_stream("Plat(): sending data: {}".format(data))
        self.serial.write(data)

    @command(dtype_in=int, doc_in='End command identifier - parameter follows')
    def End(self, val):
        """
        The CSCOMMAND_END command packet, size = 4
        The Params[] array consists of a short containing desired ramp rate
        in K/hour
        """

        HIBYTE, LOBYTE = splitBytes(val)
        data = [4, CSCOMMAND.END, HIBYTE, LOBYTE]
        self.debug_stream("End(): sending data: {}".format(data))
        self.serial.write(data)

    # Command without description on the manual. It is not used on the
    # beamlines
    # @command(dtype_in=float, doc_in='Shyt the Cryosutter for the specified '
    #                                 'length of time.')
    # def CryoShutter_Start_Auto(self, value):
    #     """
    #     The CSCOMMAND_CRYOSHUTTER_START_AUTO command packet, size = 3
    #     The Params[] array consists of a short containing the duration of the
    #     anneal time in tenths of a second
    #     """
    #     val = int(value * 10)
    #     data = [3, CSCOMMAND.CRYOSHUTTER_START_AUTO, val]
    #     self.debug_stream("CryoShutter_Start_Auto(): "
    #                       "sending data: {}".format(data))
    #     self.serial.write(data)

    @command
    def CryoShutter_Start_Man(self):
        """
        The	CSCOMMAND_CRYOSHUTTER_START_MAN command packet, size = 2
        Shut until CSCOMMAND_CRYOSHUTTER_STOP
        """
        data = [2, CSCOMMAND.CRYOSHUTTER_START_MAN]
        self.debug_stream("CryoShutter_Start_Man(): "
                          "sending data: {}".format(data))
        self.serial.write(data)

    @command
    def CryoShutter_Stop(self):
        """
        The	CSCOMMAND_CRYOSHUTTER_STOP command packet, size = 2
        Open the CryoShutter
        """
        data = [2, CSCOMMAND.CRYOSHUTTER_STOP]
        self.debug_stream("CryoShutter_Stop(): "
                          "sending data: {}".format(data))
        self.serial.write(data)

    @command(dtype_in=int, doc_in='Set status packet format: 0 old, '
                                  '1 extended')
    def Status_Format(self, args):
        """
        The CSCOMMAND_SETSTATUSFORMAT command packet, size = 3
        """

        if len(args) != 1:
            raise Exception(
                "Wrong number of arguments. Required paramters 0 or 1.")
        try:
            val = int(args[0])
            if val != 0 and val != 1:
                raise Exception()
        except Exception:
            raise Exception(
                "Wrong type of arguments. Status_Format must be an integer, "
                "0 or 1.")

        data = [3, CSCOMMAND.SETSTATUSFORMAT, val]
        self.debug_stream("Status_Format(): "
                          "sending data: {}".format(data))
        self.serial.write(data)

    # ------------------------------------------------------------------
    # ATTRIBUTES
    # ------------------------------------------------------------------

    @attribute(name='GasSetPoint', unit='K')
    def gas_set_point(self):
        self.info_stream("read_GasSetPoint")
        return self.status_packet.gas_set_point

    @attribute(name='GasTemp', unit='K')
    def gas_temp(self):
        self.info_stream("read_GasSetPoint")
        return self.status_packet.gas_temp

    @attribute(name='GasError', unit='K')
    def gas_error(self):
        self.info_stream("read_GasError")
        return self.status_packet.gas_error

    @attribute(name='RunMode', dtype=str)
    def run_mode(self):
        self.info_stream("read_RunMode")
        return self.status_packet.run_mode

    @attribute(name='Phase', dtype=str)
    def phase(self):
        self.info_stream("read_Phase")
        return self.status_packet.phase

    @attribute(name='RampRate', dtype=int)
    def ramp_rate(self):
        self.info_stream("read_RampRate")
        return self.status_packet.ramp_rate

    @attribute(name='TargetTemp')
    def target_temp(self):
        self.info_stream("read_TargetTemp")
        return self.status_packet.target_temp

    @attribute(name='EvapTemp', unit='K')
    def evap_temp(self):
        self.info_stream("read_EvapTemp")
        return self.status_packet.evap_temp

    @attribute(name='SuctTemp', unit='K')
    def suct_temp(self):
        self.info_stream("read_SuctTemp")
        return self.status_packet.suct_temp

    @attribute(name='GasFlow', unit='l/min')
    def gas_flow(self):
        self.info_stream("read_GasFlow")
        return self.status_packet.gas_flow

    @attribute(name='GasHeat', unit='%')
    def gas_heat(self):
        self.info_stream("read_GasHeat")
        return self.status_packet.gas_heat

    @attribute(name='EvapHeat', unit='%')
    def evap_heat(self):
        self.info_stream("read_EvapHeat")
        return self.status_packet.evap_heat

    @attribute(name='SuctHeat', unit='%')
    def suct_heat(self):
        self.info_stream("read_SuctHeat")
        return self.status_packet.suct_heat

    @attribute(name='LinePressure', unit='bar')
    def line_pressure(self):
        self.info_stream("read_LinePressure")
        return self.status_packet.line_pressure

    @attribute(name='Alarm', dtype=str)
    def alarm(self):
        self.info_stream("read_Alarm")
        return self.status_packet.alarm

    @attribute(name='RunTime', dtype=str)
    def run_time(self):
        self.info_stream("read_RunTime")
        result = '{}d, {}h, {}m'.format(self.status_packet.run_days,
                                        self.status_packet.run_hours,
                                        self.status_packet.run_mins)
        return result

    @attribute(name='ControllerNr', dtype=int)
    def controller_number(self):
        self.info_stream("read_ControllerNumber")
        return self.status_packet.controller_nb

    @attribute(name='SoftwareVersion', dtype=int)
    def software_version(self):
        self.info_stream("read_SoftwareVersion")
        return self.status_packet.software_version

    @attribute(name='EvapAdjust', dtype=int)
    def evap_adjust(self):
        self.info_stream("read_EvapAdjust")
        return self.status_packet.evap_adjust

    @attribute(name='TurboMode', dtype=bool)
    def turbo_mode(self):
        self.info_stream("read_TurboMode")
        flow = self.status_packet.gas_flow
        phase = self.status_packet.phase
        self.info_stream('flow: {} , phase: {}'.format(flow, phase))
        if flow > 5 and phase != "Cool":
            turbomode = True
        else:
            turbomode = False
        return turbomode

    def update_status_packet(self):
        self.status_thread_stop.clear()
        # flushing input buffer
        self.flush_input_buffer()
        # updating loop
        while not self.status_thread_stop.isSet():
            # TODO Implement check of the status format package it can be
            #  extended
            raw_data = self.serial.read(32)
            if self.serial.inWaiting() > 32:
                # if there are newer packets in the buffer, we do not process
                # and just continue
                continue
            data = list(map(int, raw_data))
            try:
                self.status_packet = StatusPacket(data)
            except Exception as e:
                self.error_stream("Error while parsing read data: %s" % e)
                self.error_stream("Flushing input buffer to start from "
                                  "the skratch")
                self.flush_input_buffer()

    def flush_input_buffer(self):
        while self.serial.inWaiting() > 0:
            self.serial.flushInput()


def main():
    OxfCryo700.run_server()


if __name__ == '__main__':
    main()


oxfcryo700/oxfordcryo.py
""" BL13-XALOC
      End station
        Elements
          Cryogenic system - 700 Series - Oxford Cryosystems
            cryo_700_series_manual.pdf
              Pages 38-39:

     Status Packets
     ---------------
     As indicated above, the controller issues status packets of a
     fixed length at 1 second intervals. The format of the status
     packet is given below.

     typedef struct {
        unsigned char Length; /* Length of this packet = 32 (bytes) */
        unsigned char Type; /* Status Packet ID = 1 */
        unsigned short GasSetPoint; /* Set Temp 100*K */
        unsigned short GasTemp; /* Gas Temp 100*K */
        signed short GasError; /* Error 100*K */
        unsigned char RunMode; /* The current 'run mode' */
        unsigned char PhaseId; /* theState.PhaseTable[0].Id */
        unsigned short RampRate; /* theState.PhaseTable[0].Temp */
        unsigned short TargetTemp; /* theState.PhaseTable[0].Temp */
        unsigned short EvapTemp; /* Evap temp, 100*K */
        unsigned short SuctTemp; /* Suct temp, 100*K */
        unsigned short Remaining; /* Time remaining in phase */
        unsigned char GasFlow; /* Gas flow, 10*l/min */
        unsigned char GasHeat; /* Gas heater, % */
        unsigned char EvapHeat; /* Evap heater, % */
        unsigned char SuctHeat; /* Suct heater, % */
        unsigned char LinePressure; /* Back pressure, 100*bar */
        unsigned char AlarmCode; /* Indicates most serious alarm condition */
        unsigned short RunTime; /* Time in minutes pump has been up */
        unsigned short ControllerNumber; /* Controller number, from ROM */
        unsigned char SoftwareVersion; /* Software version */
        unsigned char EvapAdjust; /* EvapAdjust vacuum compensation */
     } CryostreamStatus ;

     Notes
      * chars have a size of 1 byte, shorts have a size of 2 bytes.
      * All temperatures are in centi-Kelvin, i.e. 80 K is reported as 8000.
      * The RunMode member make take the following values:

     enum {
        StartUp, /* = 0: Initial transient value - run through system checks */
        StartUpFail, /* = 1: Some failure in system checks - leave results on
        screen */
        StartUpOK, /* = 2: System checks OK - awaiting Start button */
        Run, /* = 3: Gas is flowing */
        SetUp, /* = 4: Special commissioning mode */
        ShutdownOK, /* = 5: System has shut down cleanly */
        ShutdownFail /* = 6: System has shut down due to hardware error */
     };

     The PhaseId member may take the following values, whose meaning
     should be obvious from the manual. This parameter is meaningless
     unless iRunMode = Run.  Parameters of the current phase are
     stored in the RampRate, TargetTemp and Remaining members.

     enum {
        Ramp, /* = 0: Current phase is a Ramp */
        Cool, /* = 1: Current phase is a Cool */
        Plat, /* = 2: Current phase is a Plat */
        Hold, /* = 3: Current phase is a Hold */
        End, /* = 4: Current phase is an End */
        Purge, /* = 5: Current phase is a Purge */
        DeletePhase, /* = 6: Internal use only */
        LoadProgram, /* = 7: Internal use only */
        SaveProgram, /* = 8: Internal use only */
        Soak, /* = 9: Part of the Purge phase */
        Wait /* = 10: Part of Ramp/Wait */
     };

     The AlarmCode member make take the following values.

     enum {
        AlarmConditionNone, /* = 0: No alarms exist */
        AlarmConditionStopPressed, /* = 1: Stop button has been pressed */
        AlarmConditionStopCommand, /* = 2: Stop command received */
        AlarmConditionEnd, /* = 3: End phase complete */
        AlarmConditionPurge, /* = 4: Purge phase complete */
        AlarmConditionTempWarning, /* = 5: Temp error > 5 K */
        AlarmConditionHighPressure, /* = 6: Back pressure > 0.5 bar */
        AlarmConditionVacuum, /* = 7: Evaporator reduction at max */
        AlarmConditionStartUpFail, /* = 8: Self-check fail */
        AlarmConditionLowFlow, /* = 9: Gas flow < 2 l/min */
        AlarmConditionTempFail, /* = 10: Temp error > 25 K */
        AlarmConditionTempReadingError,/* = 11: Unphysical temp. reported */
        AlarmConditionSensorFail, /* = 12: Invalid ADC reading */
        AlarmConditionBrownOut, /* = 13: Degradation of power supply */
        AlarmConditionHeatsinkOverheat,/* = 14: Heat sink overheating */
        AlarmConditionPsuOverheat, /* = 15: Power supply overheating */
        AlarmConditionPowerLoss /* = 16: Power failure */
     };
"""


class StatusPacket:
    # 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
    #  27 28 29 30 31
    # L T GS  2GT 2GE R P 2RR   2TT   2ET   2ST   2R    GF GH EH SH LP AC
    # 2RT   2CN   SV EA
    Length_c_idx = 0
    Type_c_idx = 1
    GasSetPoint_s_idx = 2
    GasTemp_s_idx = 4
    GasError_s_idx = 6
    RunMode_c_idx = 8
    PhaseId_c_idx = 9
    RampRate_s_idx = 10
    TargetTemp_s_idx = 12
    EvapTemp_s_idx = 14
    SuctTemp_s_idx = 16
    Remaining_s_idx = 18
    GasFlow_c_idx = 20
    GasHeat_c_idx = 21
    EvapHeat_c_idx = 22
    SuctHeat_c_idx = 23
    LinePressure_c_idx = 24
    AlarmCode_c_idx = 25
    RunTime_s_idx = 26
    ControllerNumber_s_idx = 28
    SoftwareVersion_c_idx = 30
    EvapAdjust_c_idx = 31

    RUNMODE_CODES = ['StartUp', 'StartUpFail', 'StartUpOK', 'Run', 'SetUp',
                     'ShutdownOK', 'ShutdownFail']

    PHASE_CODES = ['Ramp', 'Cool', 'Plat', 'Hold', 'End', 'Purge',
                   'DeletePhase',
                   'LoadProgram', 'SaveProgram', 'Soak', 'Wait']

    ALARM_CODES = ['AlarmConditionNone', 'AlarmConditionStopPressed',
                   'AlarmConditionStopCommand', 'AlarmConditionEnd',
                   'AlarmConditionPurge', 'AlarmConditionTempWarning',
                   'AlarmConditionHighPressure', 'AlarmConditionVacuum',
                   'AlarmConditionStartUpFail', 'AlarmConditionLowFlow',
                   'AlarmConditionTempFail', 'AlarmConditionTempReadingError',
                   'AlarmConditionSensorFail', 'AlarmConditionBrownOut',
                   'AlarmConditionHeatsinkOverheat',
                   'AlarmConditionPsuOverheat',
                   'AlarmConditionPowerLoss',
                   'AlarmConditionRefrigeratorTooCold',
                   'AlarmConditionRefrigeratorTimedOut',
                   'AlarmConditionCryodriveNotResponding',
                   'AlarmConditionCryodriveError',
                   'AlarmConditionNoNitrogen',
                   'AlarmConditionNoHelium',
                   'AlarmConditionVacuumGauge',
                   'AlarmConditionVacuumReading',
                   'AlarmConditionRS232Error',
                   'AlarmConditionColdheadTempWarning',
                   'AlarmConditionColdheadTempError',
                   'AlarmConditionWaitForEnd',
                   'AlarmConditionDoNotOpen',
                   'AlarmConditionSensorNotDisconnected',
                   'AlarmConditionCryostatOpen',
                   'AlarmConditionCryostatOpenTimeOut',
                   'AlarmConditionHighTempWarning',
                   'AlarmConditionHighTempError'
                   ]

    def __init__(self, data):
        self.length = data[self.Length_c_idx]
        self.type = data[self.Type_c_idx]
        self.gas_set_point = self.getShort(
            data[self.GasSetPoint_s_idx:self.GasSetPoint_s_idx + 2]) / 100.
        self.gas_temp = self.getShort(
            data[self.GasTemp_s_idx:self.GasTemp_s_idx + 2]) / 100.
        self.gas_error = self.getSignedShort(
            data[self.GasError_s_idx:self.GasError_s_idx + 2]) / 100.
        self.run_mode_code = data[self.RunMode_c_idx]
        self.run_mode = self.RUNMODE_CODES[self.run_mode_code]
        self.phase_code = data[self.PhaseId_c_idx]
        self.phase = self.PHASE_CODES[self.phase_code]
        self.ramp_rate = self.getShort(
            data[self.RampRate_s_idx:self.RampRate_s_idx + 2])
        self.target_temp = self.getShort(
            data[self.TargetTemp_s_idx:self.TargetTemp_s_idx + 2]) / 100.
        self.evap_temp = self.getShort(
            data[self.EvapTemp_s_idx:self.EvapTemp_s_idx + 2]) / 100.
        self.suct_temp = self.getShort(
            data[self.SuctTemp_s_idx:self.SuctTemp_s_idx + 2]) / 100.
        self.remaining = self.getShort(
            data[self.Remaining_s_idx:self.Remaining_s_idx + 2])
        self.gas_flow = data[self.GasFlow_c_idx] / 10.
        self.gas_heat = data[self.GasHeat_c_idx]
        self.evap_heat = data[self.EvapHeat_c_idx]
        self.suct_heat = data[self.SuctHeat_c_idx]
        self.line_pressure = data[self.LinePressure_c_idx]
        self.alarm_code = data[self.AlarmCode_c_idx]
        self.alarm = self.ALARM_CODES[self.alarm_code]
        rt = data[self.RunTime_s_idx:self.RunTime_s_idx + 2]
        self.run_time = self.getShort(rt)
        self.run_days = self.run_time // (60 * 24)
        self.run_hours = (self.run_time - (self.run_days * 24 * 60)) // 60
        run_min = (self.run_days * 24 * 60) - (self.run_hours * 60)
        self.run_mins = self.run_time - run_min
        nb = data[self.ControllerNumber_s_idx:self.ControllerNumber_s_idx + 2]
        self.controller_nb = self.getShort(nb)
        self.software_version = data[self.SoftwareVersion_c_idx]
        self.evap_adjust = data[self.EvapAdjust_c_idx]

    def getShort(self, data):
        return (data[0] << 8) + data[1]

    def getSignedShort(self, data):
        """reverting two's complement if necessary"""
        short = self.getShort(data)
        # checking if value is negative
        if (short & 0b1000000000000000) >> 15:
            short = ~short
            short += 1
            short &= 0b1111111111111111
            short *= -1  # returning the negative value
        return short

    def __repr__(self):
        pretty_print = 'Status Packet:'
        pretty_print += '\nlength: {}'.format(self.length)
        pretty_print += '\ntype: {}'.format(self.type)
        pretty_print += '\ngas set point: {:.2f} ' \
                        '(K)'.format(self.gas_set_point)
        pretty_print += '\ngas temp: {:.2f} (K)'.format(self.gas_temp)
        pretty_print += '\ngas error: {:.2f} (K)'.format(self.gas_error)
        pretty_print += '\nrun mode code: {}'.format(self.run_mode_code)
        pretty_print += '\nrun mode: {}'.format(self.run_mode)
        pretty_print += '\nphase code: {}'.format(self.phase_code)
        pretty_print += '\nphase: {}'.format(self.phase)
        pretty_print += '\nramp rate: {} (K/h)'.format(self.ramp_rate)
        pretty_print += '\ntarget temp: {:.2f} (K)'.format(self.target_temp)
        pretty_print += '\nevap temp: {:.2f} (K)'.format(self.evap_temp)
        pretty_print += '\nsuct temp: {:.2f} (K)'.format(self.suct_temp)
        pretty_print += '\nremaining: {}'.format(self.remaining)
        pretty_print += '\ngas flow: {} (l/min)'.format(self.gas_flow)
        pretty_print += '\ngas heat: {} %'.format(self.gas_heat)
        pretty_print += '\nevap heat: {} %'.format(self.evap_heat)
        pretty_print += '\nsuct heat: {} %'.format(self.suct_heat)
        pretty_print += '\nline pressure: {} ' \
                        '(100*bar)'.format(self.line_pressure)
        pretty_print += '\nalarm code: %d' % self.alarm_code
        pretty_print += '\nalarm: %s' % self.alarm
        pretty_print += '\nrun time: {} (min): {}d {}h ' \
                        '{}m'.format(self.run_time, self.run_days,
                                     self.run_hours, self.run_mins)
        pretty_print += '\ncontroller number: {}'.format(self.controller_nb)
        pretty_print += '\nsoftware version: {}'.format(self.software_version)
        pretty_print += '\nevap adjust: {}'.format(self.evap_adjust)
        return pretty_print


class Struct:
    def __init__(self, **entries): self.__dict__.update(entries)


Enum = Struct
CSCOMMAND = Enum(RESTART=10, RAMP=11, PLAT=12, HOLD=13, COOL=14, END=15,
                 PURGE=16, PAUSE=17, RESUME=18, STOP=19, TURBO=20,
                 SETSTATUSFORMAT=40, CRYOSHUTTER_START_AUTO=80,
                 CRYOSHUTTER_START_MAN=81, CRYOSHUTTER_STOP=82)


def splitBytes(number):
    """splits high and low byte (two less significant bytes) of an integer,
    and returns them as chars"""
    if not isinstance(number, int):
        raise Exception(
            "splitBytes(number): Wrong function parameter. It must be an "
            "integer.")
    low = number & 0b11111111
    high = (number >> 8) & 0b11111111
    return high, low
